### PR TITLE
Implement new drop system. Full chest drop support.

### PR DESCRIPTION
**Pull request by eternalcomet**
## Description
Implment a new drop system,including chest rewards,boss chest rewards,monster drop,BGM CDs collection,etc.
The difference between the new and old drop system is that the new version use the drop_id field in lua scripts,which will be more proper.Besides,the new version united the drop system,improving the game experience greatly.
i wrote drop json files and implement a new drop system more like official.I want to unite different drop systems(chest,monster,investigation,BGM CD collection,book paper,etc.) together.

I renamed old DropSystem to DropSystemLegacy and rewrite DropSystem.

**NEED THESE FILES TO RUN:**
[Server.zip](https://github.com/Grasscutters/Grasscutter/files/10083474/Server.zip)
**Download it and and extract them into `resources/Server` folder.**
[data.zip](https://github.com/Grasscutters/Grasscutter/files/10054674/data.zip)(updated)
**Download it and extract it into data folder.(Not resources folder)**
If the pr is accepted,I will contribute these to the resource repository.

welcome to test whether the random system works well.
**Now new drop system is workable.**
Progress:
- [x] basic DropSystem
- [x] Chest drop
- [x] BGM CDs drop,broken book paper drop,etc.
- [x] boss chest drop
- [x] monster drop
- [x] investigation drop

## Type of changes

- [x] Bug fix
- [x] New feature 
- [x] Enhancement
- [ ] Documentation

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] My pull request is unique and no other pull requests have been opened for these changes
- [x] I have read the [Contributing note](https://github.com/Grasscutters/Grasscutter/blob/stable/CONTRIBUTING.md) and [Code of conduct](https://github.com/Grasscutters/Grasscutter/blob/development/CODE_OF_CONDUCT.md)
- [x] I am responsible for any copyright issues with my code if it occurs in the future.

### CODE REVIEW

**Comment by TYWIM** [comment]
How do I use it?

**Review by 909101 — APPROVED** [review]

**Comment by akatatsu27** [comment]
> How do I use it?

pay up

**Comment by eternalcomet** [comment]
> How do I use it?

clone my branch and compile it, or get compiled jar file from github actions. dowload the files mentioned above, put them in the specified folder and run the jar file.

File: src/main/java/emu/grasscutter/game/entity/gadget/GadgetChest.java [hunk 1]
```diff
@@ -13,16 +13,64 @@
...
     }
 
+    /**
+     * @return Whether we should remove the gadget.
+     */
     public boolean onInteract(Player player, GadgetInteractReq req) {
+        //If bigWorldScript enabled,use new drop system.
+        if (Grasscutter.getConfig().server.game.enableScriptInBigWorld) {
+            SceneGadget chest = getGadget().getMetaGadget();
+            Grasscutter.getLogger().info("OpenChest:chest_drop_id={},drop_tag={}", chest.chest_drop_id, chest.drop_tag);
```
Comment (KingRainbow44) [inline src/main/java/emu/grasscutter/game/entity/gadget/GadgetChest.java]: Please remove this for production, else put it under a debug check!

**Review by KingRainbow44 — CHANGES_REQUESTED** [review]
Besides the one suggestion, everything else _looks_ good! I will ask that you clean up your logs (add spaces between periods) as well as adding any logger messages to language files.

**Comment by Birdulon** [comment]
I have very mixed feelings about structuring tables which don't exist in the client as `ExcelBinOutput`. In this particular case, they come from actual excels, but up until now we have made a point of leaving `BinOutput` and `ExcelBinOutput` as purely translated client file dumps (see `QuestEncryptionKeys` being put in the root `Resources` folder).

**Comment by eternalcomet** [comment]
> I have very mixed feelings about structuring tables which don't exist in the client as `ExcelBinOutput`. In this particular case, they come from actual excels, but up until now we have made a point of leaving `BinOutput` and `ExcelBinOutput` as purely translated client file dumps (see `QuestEncryptionKeys` being put in the root `Resources` folder).

maybe we can talk about it on discord?

**Comment by eternalcomet** [comment]
now investigation feature is available but still have bugs.
after investigating, the place is still sparkling but uninteractable. guess it's because of improper GadgetInteractRsp.
pic(386)
pic(387)
pic(388)

**Comment by TYWIM** [comment]
I came across something interesting, some people separated Grasscutters, the private server function is almost the same as the official version, you can also do tasks, do they use the official server to do it?So how do you make it?




------------------&nbsp;原始邮件&nbsp;------------------
发件人: ***@***.***&gt;; 
发送时间: 2022年11月23日(星期三) 中午11:54
收件人: ***@***.***&gt;; 
抄送: "✨ 小🏆 桐🌸乖乖💝 ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [Grasscutters/Grasscutter] Implement new drop system. Full chest drop support. (PR #1951)






现在调查功能是可用的，但仍然有错误。
经过调查，这个地方仍然闪闪发光，但无法互动。估计是因为GadgetInteractRsp不正确。




—
直接回复此邮件，在GitHub上查看，或取消订阅.
***@***.***与&gt;.

**Comment by eternalcomet** [comment]
now fixed.
pay attention to the isOneOff field of SceneGadget.java.
this field indicates whether the gadget should disappear permanently.
For example, if isOneOff=true, like most chests, it will disappear permanently after interacted.
If isOneOff=false, like investigation points, it will disappear temporarily, and appear again in next big world resource refresh routine.
I add an comment on the code.

**Comment by eternalcomet** [comment]
@TYWIM  Should not talk about this here. The server whose function is almost the same as the official version has nothing to do with Grasscutter.

**Comment by TYWIM** [comment]
You're absolutely right, thank you for your help


------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "Grasscutters/Grasscutter"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年11月23日(星期三) 中午12:11
***@***.***&gt;;
抄送:&nbsp;"✨ 小🏆 桐🌸乖乖💝 ***@***.******@***.***&gt;;
主题:&nbsp;Re: [Grasscutters/Grasscutter] Implement new drop system. Full chest drop support. (PR #1951)





 
@TYWIM  Should not talk about this here. The server whose function is almost the same as the official version has nothing to do with Grasscutter.

**Comment by yuanshengaoshou114** [comment]
Is your modified version based on Hartie95 no longer maintained?

**Comment by eternalcomet** [comment]
@keqing114514 Yes. Hartie95's branch is a fork of this repo, mainly focus on quests fix. Hartie95 suggested me to contribute to the main repo so I did that. Hartie95 will merge this branch into his branch in the future. If you like my pr, you can merge my branch into yours. Welcome to test and feedback.

之前我的分支是基于development和Hartie95的quests分支。但是他的分支主要专注于修复任务功能，所以他建议我贡献给主仓库。如果你想使用我的分支，可以把它merge进你fork的quests分支就行。欢迎进行测试并提供反馈。

**Comment by yuanshengaoshou114** [comment]
> @keqing114514 Yes. Hartie95's branch is a fork of this repo, mainly focus on quests fix. Hartie95 suggested me to contribute to the main repo so I did that. Hartie95 will merge this branch into his branch in the future. If you like my pr, you can merge my branch into yours. Welcome to test and feedback.
> 
> 之前我的分支是基于development和Hartie95的quests分支。但是他的分支主要专注于修复任务功能，所以他建议我贡献给主仓库。如果你想使用我的分支，可以把它merge进你fork的quests分支就行。欢迎进行测试并提供反馈。

Well, although I want to use it, I can't resolve the conflict. Thanks for your help

**Comment by eternalcomet** [comment]
@keqing114514 Here is the branch you want. I updated the new commits and solved the conflicts. Now it's based on both Hartie's and main repo's code. [https://github.com/eternalcomet/ys/tree/drop-table](https://github.com/eternalcomet/ys/tree/drop-table)

**Comment by yuanshengaoshou114** [comment]
> @keqing114514 Here is the branch you want. I updated the new commits and solved the conflicts. Now it's based on both Hartie's and main repo's code. https://github.com/eternalcomet/ys/tree/drop-table

Cannot merge, perhaps because I use hidden branches, and task branches can be merged... But now I can only use hidden branches, can you modify them?

**Comment by TYWIM** [comment]
Sorry, I've been working on a personal website these days, I didn't pay attention to it, and now how to use it




------------------&nbsp;原始邮件&nbsp;------------------
发件人: ***@***.***&gt;; 
发送时间: 2022年11月23日(星期三) 晚上8:08
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [Grasscutters/Grasscutter] Implement new drop system. Full chest drop support. (PR #1951)






@keqing114514这是你要的树枝。我更新了新提交并解决了冲突。现在它基于Hartie和主回购的代码。https://github.com/eternalcomet/ys/tree/drop-table

无法合并，可能是因为我使用了隐藏分支，而任务分支可以合并...但是现在我只能用隐藏的分支，可以修改吗？

—
直接回复此邮件，在GitHub上查看，或取消订阅.
***@***.***与&gt;.

**Comment by GanyusLeftHorn** [comment]
Very cool work, thank you! 

There is one thing here that I would strongly suggest changing. The drop tables should directly be part of this repository and go in `src/main/resources/defaults/data/`. So far, `resources/` was always reserved for data that was directly extracted from the client, and I would not change that.

**Comment by akbaryahya** [comment]
in some cases if computer can't read chinese name (drop_tag), try running it with `java -Dfile.encoding=UTF-8 -jar grasscutter.jar`

**Comment by eternalcomet** [comment]
@GanyusLeftHorn Thanks for your advice. @Birdulon previously asked me to put these Excel-like file into `resources/Server` folder. So I changed that.

**Comment by yuanshengaoshou114** [comment]
> @keqing114514 Here is the branch you want. I updated the new commits and solved the conflicts. Now it's based on both Hartie's and main repo's code. https://github.com/eternalcomet/ys/tree/drop-table

I am using https://github.com/Hartie95/Grasscutter/tree/hidden , cannot communicate with https://github.com/eternalcomet/ys/tree/drop-table Merge, can you modify it? Hidden modified the file. I can't use the quest version @eternalcomet

**Comment by keqing114514** [comment]
> @keqing114514 Here is the branch you want. I updated the new commits and solved the conflicts. Now it's based on both Hartie's and main repo's code. https://github.com/eternalcomet/ys/tree/drop-table

Can you help me? riding crop

**Comment by eternalcomet** [comment]
@GanyusLeftHorn

**Review by GanyusLeftHorn — CHANGES_REQUESTED** [review]
Okay, I've finally gotten around to having a look at this. The system itself seems to work really well - great work! However, there are a couple problems/open questions we need to address before we can merge this:

1.) I actually had to massage this a little to get it working (I had to compile the protos for DropHintNotify and ClosedItemNotify myself, otherwise I would get compile errors). Similarly, when trying to merge this onto the current development branch, there are merge conflicts. Can you please get this into a mergeable state?

2.) The files in `resources/Server` should be added to the resource repositories.

3.) The files in `data` should be added to `src/main/resources/defaults/data`, so they are distributed with GC. I assume those a handwritten, not extracted from the client, correct?

**Comment by GanyusLeftHorn** [comment]
Oh, one more think I noticed: Investigation points are giving me primogems instead of the usual drops.

**Comment by eternalcomet** [comment]
> 

which investigation point?

**Comment by eternalcomet** [comment]
> Okay, I've finally gotten around to having a look at this. The system itself seems to work really well - great work! However, there are a couple problems/open questions we need to address before we can merge this:
> 
> 1.) I actually had to massage this a little to get it working (I had to compile the protos for DropHintNotify and ClosedItemNotify myself, otherwise I would get compile errors). Similarly, when trying to merge this onto the current development branch, there are merge conflicts. Can you please get this into a mergeable state?
> 
> 2.) The files in `resources/Server` should be added to the resource repositories.
> 
> 3.) The files in `data` should be added to `src/main/resources/defaults/data`, so they are distributed with GC. I assume those a handwritten, not extracted from the client, correct?

thanks for your review.
1) I managed to solved the conflict.
2) now the resource repository only provide a zipped file, I have no idea how to contribute to it. Or do you know which resource repository can I contribute to?
3) No. these files are converted from `txt` files so I don't think it's proper to distribute it with GC. Maybe I should place them in the resource. Do you have any idea?

**Comment by GanyusLeftHorn** [comment]
Great, thank you for resolving the conflicts! To contribute to the resources repo, I suggest you just submit a PR to it (on Crepe git - you know that one, I assume?).

As for the files, yes, if they are converted server files, they should go in `resources`, not `data`. That is only meant for files we actually made ourselves, and that get auto-populated by GC. I don't want to force users to manually put files into `data` as well now.

**Comment by StartForKiller** [comment]
> now fixed. pay attention to the isOneOff field of SceneGadget.java. this field indicates whether the gadget should disappear permanently. For example, if isOneOff=true, like most chests, it will disappear permanently after interacted. If isOneOff=false, like investigation points, it will disappear temporarily, and appear again in next big world resource refresh routine. I add an comment on the code.

This is correct in most cases but not in all of them, for example, if a chest isOneOff is true but permanent flag is not true, the gadget state is not saved and the entity is respawned on the next load/unload group thing. Most of the cases isOneOff is set to true at the same time permanent is set to true, but there are cases where not. I'm working on dynamic group loading and adding support for saving and loading groups

**Comment by KingRainbow44** [comment]
closing due to inactivity; feel free to reopen/create a new PR with the requested changes if you wish

**Pull request closed without merging**